### PR TITLE
37549 encryption key rotation validation tests

### PR DESCRIPTION
**Pull request by jakefhyde**
## Issue:  

https://github.com/rancher/rancher/issues/37549 

Implemented encryption key rotation validaiton tests for both a fresh install, and a cluster with 10k secrets. Creating 10k secrets takes a long time, as does reencrypting them (>1 hour in total) so any suggestions for speeding things up are welcome.

## Testing

I've tested the following cases based of the latest 1.24 releases for rke2 & k3s, which succeeded using a custom branch I have used to fix planner thrashing. We should be okay to merge this PR beforehand however, since these tests aren't automated. 

configuration | status
-- | --
1 node all roles | ✔️
1 etcd, 1 cp, 1 worker | ✔️
3 nodes all roles | ✔️
3 etcd, 2 cp, 3 worker | ✔️
3 etcd, 3 cp, 3 worker | ✔️
3 etcd, 5 cp, 3 worker | ✔️

### CODE REVIEW

File: tests/framework/extensions/machinepools/machinepools.go
Comment (annablender) [inline tests/framework/extensions/machinepools/machinepools.go:76]: Looks like lines 76-92 are duplicate code. Why is it commented out? Can it be removed?

File: tests/v2/validation/provisioning/rke2/rke2_encryption_key_rotation_test.go
Comment (annablender) [inline tests/v2/validation/provisioning/rke2/rke2_encryption_key_rotation_test.go:65]: Will provider ever be nil?

File: tests/v2/validation/provisioning/rke2/rke2_encryption_key_rotation_test.go
Comment (annablender) [inline tests/v2/validation/provisioning/rke2/rke2_encryption_key_rotation_test.go:75]: Namespace is never passed into the func. It's a field of the `RKE2EncryptionKeyROtationTestSuite` struct, shouldn't it be `r.namespace?` since I'm assuming you've smoke tested this and it works, I'm probably missing something about Go here

Comment (jakefhyde) [reply tests/v2/validation/provisioning/rke2/rke2_encryption_key_rotation_test.go:75]: No you're right it probably needs to be updated to `r.namespace`.

File: tests/framework/extensions/machinepools/machinepools.go
Comment (jakefhyde) [reply tests/framework/extensions/machinepools/machinepools.go:76]: It's a comment explaining how it works. Although the comment does seem out of date, I'll update.

File: tests/v2/validation/provisioning/rke2/rke2_encryption_key_rotation_test.go
Comment (jakefhyde) [reply tests/v2/validation/provisioning/rke2/rke2_encryption_key_rotation_test.go:75]: I lied, `namespace` is defined in `tests/v2/validation/provisioning/rke2/config.go` as `fleet-default`, which makes sense.

File: tests/v2/validation/provisioning/rke2/rke2_encryption_key_rotation_test.go
Comment (jakefhyde) [inline tests/v2/validation/provisioning/rke2/rke2_encryption_key_rotation_test.go:124]: @igomez06 Any idea why this sin't working? It always errors with "the server could not find the requested resource"

Comment (igomez06) [reply tests/v2/validation/provisioning/rke2/rke2_encryption_key_rotation_test.go:124]: It's because you are passing the cluster name rather than the clusterID

File: pkg/codegen/main.go
Comment (igomez06) [inline pkg/codegen/main.go:110]: This is actually generating the k8's rke client not the rancher one. So this client would be hitting the Kubernetes api endpoint rather than the rancher. One of the last thing's Jacob had suggested was to use generate the client similar to how the management.cattle.io clients are generated using the Norman code generator. I did it for the provisioning.cattle.io client in this PR.

https://github.com/rancher/rancher/pull/38281

**Comment by igomez06** [comment]
I think this PR: https://github.com/rancher/rancher/pull/38281, needs to be approved and merged before merging this one. As the provisioning client has been updated.

File: tests/v2/validation/provisioning/rke2/rke2_encryption_key_rotation_test.go
Comment (jakefhyde) [reply tests/v2/validation/provisioning/rke2/rke2_encryption_key_rotation_test.go:124]: Oh, duh

File: tests/framework/clients/rancher/rke/client.go
Comment (jakefhyde) [reply tests/framework/clients/rancher/rke/client.go:20]: AFAIK no, I'm only using it to watch a status field on the controlplane.

File: tests/framework/extensions/machinepools/machinepools.go
Comment (jakefhyde) [reply tests/framework/extensions/machinepools/machinepools.go:76]: I have, as a side note I like the `machinepools.NodeRools` mechanism it's really easy to work with and it's really easy to structure the json for in `CATTLE_TEST_CONFIG`.

**Comment by jakefhyde** [comment]
This PR depends on https://github.com/rancher/rancher/pull/38796, so you may see some changes to `encryptionkeyrotation.go`, for the purposes of this PR just ignore them.

**Review by igomez06 — COMMENTED** [review]
So my question is Jake why are you using the kubeapi to create these v1 resources rather than the Steve client? Since this in validation tests, we want to purely mimic the UI ie. use the Steve client to create v1 resources, not the controllers. Can you see how we did in the cert_rotation test in tests/v2/validation/provisioning/rke2?

**Comment by jakefhyde** [comment]
@igomez06 We are using the kubeapi to create secrets as part of the tests, because the creation of secrets is required to test whether or not encryption key rotation works with clusters with large numbers of secrets, the creation of these resources is not being tested here. We are still using steve to create/update the provisioning cluster.

**Comment by igomez06** [comment]
> @igomez06 We are using the kubeapi to create secrets as part of the tests, because the creation of secrets is required to test whether or not encryption key rotation works with clusters with large numbers of secrets, the creation of these resources is not being tested here. We are still using steve to create/update the provisioning cluster.

@jakefhyde Ah ok, so let me make sure I understand. The 10k secrets is a setup/perquisite of the main test being that  encryption key rotation works with clusters. The secrets part make sense, but you using the controller for updating the provisioned cluster that I pointed out. Shouldn't we use the Steve client to update the cluster, as that is part of the test?

**Comment by jakefhyde** [comment]
@igomez06 Updated, thanks for being patient with this review, it's taking a bit longer than expected but the issues with encryption key rotation for a large # of secrets have been determined so PRs are going to go up soon.

**Comment by jakefhyde** [comment]
Still debugging timeouts

File: tests/framework/extensions/kubeapi/secrets/create.go
Comment (igomez06) [inline tests/framework/extensions/kubeapi/secrets/create.go:15]: What's the difference between this func, and other `CreateSecret` func?

**Review by igomez06 — COMMENTED** [review]
Looks great Jake just had one question

File: tests/framework/extensions/kubeapi/secrets/create.go
Comment (jakefhyde) [reply tests/framework/extensions/kubeapi/secrets/create.go:15]: The difference being that this function just wraps the functionality for creating the dynamic client for the cluster, so that in my encryption key rotation tests I don't have to recreate the client in the creation loop. I don't know if it creates a measurable performance impact, but it takes like an hour to generate 10k secrets so figured it was probably better to do it this way.

**Review by igomez06 — APPROVED** [review]
Looks good. We're holding off on test merges until our HA upgrade PR goes in. This requires v2.6 backport correct?

File: tests/v2/validation/provisioning/rke2/encryption_key_rotation_test.go
Comment (bfbachmann) [inline tests/v2/validation/provisioning/rke2/encryption_key_rotation_test.go:41]: Sorry, I lack context here so I'm just curious: why 10k?

File: tests/v2/validation/provisioning/rke2/encryption_key_rotation_test.go
Comment (bfbachmann) [inline tests/v2/validation/provisioning/rke2/encryption_key_rotation_test.go:148]: I see we're passing `timeout` here, but how come you're only using it when watching `RKEControlPlanes` and using `defaults.WatchTimeoutSeconds` when watching `Clusters`? 

Also, this is probably a nitpick, but should `timeout` here be of type `time.Duration` rather than `int64`?

File: tests/v2/validation/provisioning/rke2/encryption_key_rotation_test.go
Comment (bfbachmann) [inline tests/v2/validation/provisioning/rke2/encryption_key_rotation_test.go:110]: Nit: You're creating this variable but it's only used in one place, which happens to be on the next line.

File: tests/v2/validation/provisioning/rke2/encryption_key_rotation_test.go
Comment (bfbachmann) [inline tests/v2/validation/provisioning/rke2/encryption_key_rotation_test.go:76]: Sorry if this is a stupid question, but what does this `initial` test label refer to? Are we planning on creating more test sub-suites alongside this `initial` one?

File: tests/v2/validation/provisioning/rke2/encryption_key_rotation_test.go
Comment (jakefhyde) [reply tests/v2/validation/provisioning/rke2/encryption_key_rotation_test.go:41]: context: https://github.com/rancher/rancher/issues/40202

File: tests/v2/validation/provisioning/rke2/encryption_key_rotation_test.go
Comment (jakefhyde) [reply tests/v2/validation/provisioning/rke2/encryption_key_rotation_test.go:148]: From apimachinery:
```
	TimeoutSeconds *int64 `json:"timeoutSeconds,omitempty" protobuf:"varint,5,opt,name=timeoutSeconds"`
```

File: tests/v2/validation/provisioning/rke2/encryption_key_rotation_test.go
Comment (jakefhyde) [reply tests/v2/validation/provisioning/rke2/encryption_key_rotation_test.go:76]: I was originally going to separate the provisioning and the rotation, but scrapped that idea, so this "initial" test case can be unwrapped, good catch.

File: tests/v2/validation/provisioning/rke2/encryption_key_rotation_test.go
Comment (jakefhyde) [reply tests/v2/validation/provisioning/rke2/encryption_key_rotation_test.go:110]: Just copying the logic here: https://github.com/rancher/rancher/blob/10f2cc093c4864a694981338c034acfdd20b6548/tests/v2/validation/provisioning/rke2/custom_cluster.go#L81-L82

Now that I think about it, your comment is probably the more correct way to do it.

File: tests/framework/extensions/kubeapi/resource.go
Comment (igomez06) [inline tests/framework/extensions/kubeapi/resource.go:9]: Can you add a comment here for our generated go docs?

File: tests/v2/validation/provisioning/k3s/encryption_key_rotation_test.go
Comment (igomez06) [inline tests/v2/validation/provisioning/k3s/encryption_key_rotation_test.go:187]: If provisioning is not considered a part of the test would it make sense to have it in its own test package and receive the cluster ID like we do in other tests?
ex) https://github.com/rancher/rancher/blob/a05de31fccb10059447c169f28dcc2068982a6f0/tests/v2/validation/charts/gatekeeper_test.go#L42-L47

Comment (igomez06) [reply tests/v2/validation/provisioning/k3s/encryption_key_rotation_test.go:187]: This way you don't have to duplicate test for k3s as well.

File: tests/v2/validation/provisioning/rke2/encryption_key_rotation_test.go
Comment (igomez06) [inline tests/v2/validation/provisioning/rke2/encryption_key_rotation_test.go:186]: same here

**Review by igomez06 — COMMENTED** [review]
Looks great, would probably make sense to move this test to its own package, since it's not a provisioning test.

**Review by igomez06 — APPROVED** [review]

**Review by bfbachmann — APPROVED** [review]

**Pull request merged**
